### DISCUSSION TITLE
Using furrr

### DISCUSSION THREAD

**Discussion by 1beb** [ropensci/targets#359] [Help] [answered]
Are there any examples available that use furrr?

**Comment by wlandau** [comment, marked as answer]
Not that I am aware, but the principle is straightforward if you use ordinary `tar_make()`.

```r
# _targets.R file
library(targets)
library(future.callr)
library(furrr)
plan(callr) # Declare the plan here for tar_make()
tar_target(target_that_uses_furrr, future_map(...))
```

If you are also parallelizing the work with `tar_make_future()`, you would declare two futures in `_targets.R`: one to parallelize among targets, and another to parallelize *within* targets (e.g. `furrr`).

```r
# _targets.R file
library(targets)
library(future.callr)
library(furrr)
list(plan(callr), plan(callr))
tar_target(target_that_uses_furrr, future_map(...))
```

For `tar_make_clustermq()`, each future plan would need to be reconstructed inside the target itself.

```r
# _targets.R file
library(targets)
library(future.callr)
library(furrr)
tar_target(target_that_uses_furrr, {
  plan(callr)
  future_map(...)
})
```

**Reply by 1beb** [reply]
I'm struggling with this a little bit because I'm having trouble identifying at what level I should run in parallel and how to interact the requirements for next steps. 

Here's a slimmed down example of what I do in a script right now: 

```
future_map(ids, a_function_that_creates_a_file_per_id) # step 1
future_map(ids, a_function_that_parses_those_files) # step 2, requires data from step 1
basic_processes() # step 3, requires data from 2
extra_long_single_core_process # step 3, requires data from 2
future_map(ids, a_summarization_function) # step 4, requires data from 2
```

I have about 500 ids. I'm having trouble understanding how I can associate the data files with the future process in the target.

**Reply by wlandau** [reply]
Sequential `future_map()`'s can be inefficient because all tasks in an upstream stage need to finish before any task in a downstream stage can begin. Inversely, `tar_make_future()` and `tar_make_clustermq()` are more efficient because they allow downstream targets to begin whenever their upstream dependencies are ready and a worker is available (see the [animation here](https://books.ropensci.org/targets/hpc.html#clustermq)). 

500 ids will probably max out the parallelism capabilities of your local laptop or cluster no matter how you slice it, so it is best to choose one or the other. I recommend going all in on `targets` parallelism and dropping `furrr` entirely. It's much more convenient because `targets` does all the orchestration for you. 

The one thing is that parallelizing over 500 tasks may be a bit much as far as overhead is concerned, so you might consider [batching](https://books.ropensci.org/targets/dynamic.html#batching) the work into [strategically large](https://books.ropensci.org/targets/practices.html#how-to-define-good-targets) groups of tasks to strike a good balance. Target factories like [`tar_rep()`](https://docs.ropensci.org/tarchetypes/reference/tar_rep.html) from [`tarchetypes`](https://docs.ropensci.org/tarchetypes/) perform automatic batching in some [dynamic branching](https://books.ropensci.org/targets/dynamic.html) scenarios.

**Reply by 1beb** [reply]
Ah! I think I see it now. I wasn't connecting the singularly named variables with map. This output made it clearer: 

![image](https://user-images.githubusercontent.com/421859/112086435-3821f880-8b5a-11eb-8dd1-f0ba92a2cfbd.png)

Thank you!

**Reply by george-wood** [reply]
I have a follow-up question about how to parallelize both among targets and within targets using `furrr`. I followed the below approach.

> If you are also parallelizing the work with `tar_make_future()`, you would declare two futures in `_targets.R`: one to parallelize among targets, and another to parallelize _within_ targets (e.g. `furrr`).
> 
> ```r
> # _targets.R file
> library(targets)
> library(future.callr)
> library(furrr)
> list(plan(callr), plan(callr))
> tar_target(target_that_uses_furrr, future_map(...))
> ```

Using `tar_make_future()`, here is what I tried:

```
tar_script({
  library(targets)
  library(future.callr)
  library(furrr)

  list(plan(callr), plan(callr))

  list(
    tar_target(fast, future_map(c(1, 3), \(x) Sys.sleep(x))),
    tar_target(slow, future_map(1:3, \(x) Sys.sleep(x)))
  )
})
```

The among target parallelization worked as expected (both targets started at approximately the same time). However, the within-target parallelization ran sequentially. The fast target takes just over 3 seconds to build and the slow target takes just over 6 seconds. My expectation was that they would both build in just over 3 seconds.

I tried this outside of `targets`, using only `future`, and both ran in just over 3 seconds. I've also tried setting a specific amount of workers, e.g. `plan(multisession, workers = 6)`. No success, unfortunately.

I'm wondering if the problem lies in the declaration of two futures, e.g. `list(plan(callr), plan(callr))`, but I've run out of ideas. Any suggestions or ideas would be much appreciated. Thanks!

**Reply by wlandau** [reply]
I think you may need to call `plan(list(callr, callr))` rather than `list(plan(callr), plan(callr))` (https://cran.r-project.org/web/packages/future/vignettes/future-3-topologies.html).

**Reply by george-wood** [reply]
Thank you!

The below works as expected when using `tar_make_future(workers = availableCores())`:

```
tar_script({
  library(targets)
  library(future.callr)
  library(furrr)

  plan(list(tweak(callr, workers = 2), tweak(callr, workers = 3)))

  list(
    tar_target(fast, future_map(c(1, 3), \(x) Sys.sleep(x))),
    tar_target(slow, future_map(1:3, \(x) Sys.sleep(x)))
  )
})
```

When using `plan(list(callr, callr))`, the among targets parallelization worked but not the within target. I am guessing that this is caused by the first `callr` declaration taking up all `availableCores()` due to the default `callr(workers = availableCores(), ...)`. Setting a specified number of workers using `tweak` seems to get around this.
